machinery/gateway_api_topology.go

```diff
@@ -170,6 +256,110 @@ func NewGatewayAPITopology(options ...GatewayAPITopologyOptionsFunc) *Topology {
+				LinkTCPRouteRuleToServicePortFunc(tcpRouteRules),   // TCPRouteRule -> ServicePort
+				LinkTCPRouteRuleToServiceFunc(tcpRouteRules, true), // TCPRoute -> service
+			))
+		} else {
+			opts = append(opts, WithLinks(LinkTCPRouteRuleToServiceFunc(tcpRouteRules, false))) // TCPRouteRule -> Service
+		}
+	} else {
+		if o.ExpandServicePorts {
+			opts = append(opts, WithLinks(
+				LinkTCPRouteToServicePortFunc(o.TCPRoutes),   // TCPRoute -> ServicePort
+				LinkTCPRouteToServiceFunc(o.TCPRoutes, true), // TCPRoute -> Service
+			))
+		} else {
+			opts = append(opts, WithLinks(LinkTCPRouteToServiceFunc(o.TCPRoutes, false))) // TCPRoute -> Service
+		}
+	}
+
+	if o.ExpandTLSRouteRules {
+		tlsRouteRules := lo.FlatMap(o.TLSRoutes, TLSRouteRulesFromTLSRouteFunc)
+		opts = append(opts, WithTargetables(tlsRouteRules...))
+		opts = append(opts, WithLinks(LinkTLSRouteToTLSRouteRuleFunc()))
+
+		if o.ExpandServicePorts {
+			servicePorts := lo.FlatMap(o.Services, ServicePortsFromBackendFunc)
+			opts = append(opts, WithTargetables(servicePorts...))
+			opts = append(opts, WithLinks(
+				LinkTLSRouteRuleToServicePortFunc(tlsRouteRules),   // TLSRouteRule -> ServicePort
+				LinkTLSRouteRuleToServiceFunc(tlsRouteRules, true), // TLSRouteRule -> Service
+			))
+		} else {
+			opts = append(opts, WithLinks(LinkTLSRouteToServiceFunc(o.TLSRoutes, false))) // TLSRoute -> Service
+		}
+	} else {
+		if o.ExpandServicePorts {
+			opts = append(opts, WithLinks(
+				LinkTLSRouteToServicePortFunc(o.TLSRoutes),   // TLSRoute -> ServicePort
+				LinkTLSRouteToServiceFunc(o.TLSRoutes, true), // TLSRoute -> Service
+			))
+		} else {
+			opts = append(opts, WithLinks(LinkTLSRouteToServiceFunc(o.TLSRoutes, false))) // TLSRoute -> Service
+		}
+	}
+
+	if o.ExpandUDPRouteRules {
+		udpRouteRules := lo.FlatMap(o.UDPRoutes, UDPRouteRulesFromUDPRouteFunc)
+		opts = append(opts, WithTargetables(udpRouteRules...))
+		opts = append(opts, WithLinks(LinkUDPRouteToUDPRouteRuleFunc()))
+
+		if o.ExpandServicePorts {
+			servicePorts := lo.FlatMap(o.Services, ServicePortsFromBackendFunc)
+			opts = append(opts, WithTargetables(servicePorts...))
+			opts = append(opts, WithLinks(
+				LinkUDPRouteRuleToServicePortFunc(udpRouteRules),   // UDPRouteRule -> ServicePort
+				LinkUDPRouteRuleToServiceFunc(udpRouteRules, true), // UDPRouteRule -> Service
+			))
+		} else {
+			opts = append(opts, WithLinks(LinkUDPRouteToServiceFunc(o.UDPRoutes, false))) // UDPRoute -> Service
+		}
+	} else {
+		if o.ExpandServicePorts {
+			opts = append(opts, WithLinks(
+				LinkUDPRouteToServicePortFunc(o.UDPRoutes),   // UDPRoute -> ServicePort
+				LinkUDPRouteToServiceFunc(o.UDPRoutes, true), // UDPRoute -> Service
+			))
+		} else {
+			opts = append(opts, WithLinks(LinkUDPRouteToServiceFunc(o.UDPRoutes, false))) // UDPRoute -> Service
+		}
+	}
+
 	if o.ExpandServicePorts {
 		opts = append(opts, WithLinks(LinkServiceToServicePortFunc())) // Service -> ServicePort
```

Comment:
Related to https://github.com/Kuadrant/policy-machinery/pull/16/files#r1719696707

```suggestion
		servicePorts := lo.FlatMap(o.Services, ServicePortsFromBackendFunc)
		opts = append(opts, WithTargetables(servicePorts...))
		opts = append(opts, WithLinks(LinkServiceToServicePortFunc())) // Service -> ServicePort
```